Add dockerfile and test so we can test simple remoting behaviors

JamesWTruher opened the pull request:
this docker file creates an image which can be used to test remoting with containers.
We start with the microsoft/windowsservercore image and then set up the system for
testing by adding a user, setting up configuration for basic auth, installing
PowerShell Core, and creating a configuration endpoint for PSCore.

The tests are very simple; it retrieve $PSVersionTable.PSVersion in the following
connections:
   Full -> Full
   Full -> Core
   Core -> Full
   Core -> Core

This is the starting point for our remoting testing using containers

JamesWTruher committed:
Add dockerfile and test so we can test simple remoting behaviors

this docker file creates an image which can be used to test remoting with containers.
We start with the microsoft/windowsservercore image and then set up the system for
testing by adding a user, setting up configuration for basic auth, installing
PowerShell Core, and creating a configuration endpoint for PSCore.

The tests are very simple; it retrieve $PSVersionTable.PSVersion in the following
connections:
   Full -> Full
   Full -> Core
   Core -> Full
   Core -> Core

mirichmo commented on test/docker/networktest/DockerRemoting.Tests.ps1:
```diff
@@ -0,0 +1,60 @@
+            $pending = $false
+        }
+        # this is all I could figure out how to set up a container which
+        # sticks around for a while
```
What about the container lifecycle commands? `docker container start`, `docker exec`, `docker container stop`, etc? `docker ps` and `docker container list` both show running containers.

Using those commands should prevent the need to use start-sleep.

mirichmo commented on test/docker/networktest/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+
+# the source msi should change on a daily basis
+# the destination should not change
+ADD PSCore.msi /PSCore.msi
```
How does this file get to the Dockerfile "root" directory? Shouldn't it be downloaded from Myget?

JamesWTruher replied:
this is done with the New-DockerTestBuild.ps1 script. I broke it out to it's own function.

mirichmo commented on test/docker/networktest/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+# the msipath (PSCore.msi) below will need to change based on the daily package
+# set up for basic auth
+# install-powershellremoting will restart winrm service
+RUN net user testuser 11aa!!AA /add; `
```
As long as you are using PowerShell...
New-LocalUser -Name testuser -Password (ConvertTo-SecureString 11aa!!AA -AsPlainText -Force)

mirichmo commented on test/docker/networktest/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+# set up for basic auth
+# install-powershellremoting will restart winrm service
+RUN net user testuser 11aa!!AA /add; `
+    net localgroup administrators testuser /add; `
```
Add-LocalGroupMember -Group Administrators -Member testuser

mirichmo commented on test/docker/networktest/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+    Start-Process -FilePath msiexec.exe -ArgumentList '-qn', `
+    '-i c:\PSCore.msi','-log c:\PSCore-install.log','-norestart' -wait ; `
+    $psexec = get-item -path ${ENV:ProgramFiles}/powershell/*/powershell.exe; `
+    $corehome = $psexec.directory.fullname; `
```
You should delete the MSI within this layer to save space

JamesWTruher replied:
heh - 23mb over a 10GB image - not sure it's a big saving, but ok.

JamesWTruher committed:
Add new file to bootstrap the docker image

fix the tests to be more resilient to changes in the version of the PSCore package

JamesWTruher committed:
update script to use local user cmdlets rather than net.exe

also remove pscore.msi at the end of the image build process to save space

JamesWTruher commented on test/docker/networktest/New-DockerTestBuild.ps1:
```diff
@@ -0,0 +1,294 @@
+    )
+
+    $headers = @{
+      'Authorization' = "Bearer {0}" -f $Constants.Token
```
fixed

JamesWTruher commented on test/docker/networktest/New-DockerTestBuild.ps1:
```diff
@@ -0,0 +1,294 @@
+    param(
+        [ValidateNotNullOrEmpty()]
+        [string]$ExtensionBranch = 'master',
+        [string]$Status = "success",
```
fixed

JamesWTruher commented on test/docker/networktest/New-DockerTestBuild.ps1:
```diff
@@ -0,0 +1,294 @@
+    [datetime]$start = ((get-date).AddDays(-${Days}))
+    $results = Get-AppVeyorBuildRange -Start $start -LastBuildId $null -ExtensionBranch $ExtensionBranch
+    $results
+    #$results |?{$_.status -match "success"} | sort-object StartedDay | Select -First $Last
```
fixed

JamesWTruher commented on test/docker/networktest/New-DockerTestBuild.ps1:
```diff
@@ -0,0 +1,294 @@
+    $URI = "{0}/projects/{1}/{2}/history?recordsNumber={3}{4}&branch{5}" -f $Constants.ApiUrl,$Constants.AccountName,
+        $Constants.ProjectName,$Records,$startBuildIdString,$ExtensionBranch
+    $project = Invoke-RestMethod -Method Get -Uri $URI
+    # "$($Constants.ApiUrl)/projects/$($Constants.AccountName)/$($Constants.ProjectName)/history?recordsNumber=$Records$startBuildIdString&branch=$ExtensionBranch"
```
fixed

JamesWTruher commented on test/docker/networktest/New-DockerTestBuild.ps1:
```diff
@@ -0,0 +1,294 @@
+}
+# Check to see if we already have an image, and if so
+# delete it if -Force was used, otherwise throw and exit
+$TestImage = docker images remotetestimage --format '{{.Repository}}'
```
oversight - fixed

JamesWTruher committed:
clean up commented lines and remove an unused parameter from one of the functions

also use constants more consistently

mirichmo commented on test/docker/networktest/New-DockerTestBuild.ps1:
```diff
@@ -0,0 +1,292 @@
+    }
+}
+# check again - there could be some permission problems
+$TestImage = docker images remotetestimage --format '{{.Repository}}'
```
Here too

JamesWTruher committed:
remove reference to docker image name by string and use variable instead.

JamesWTruher commented:
i think i have them all

JamesWTruher commented:
thanks!
